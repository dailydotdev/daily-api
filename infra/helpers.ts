import * as pulumi from '@pulumi/pulumi';
import * as gcp from '@pulumi/gcp';
import * as inputs from '@pulumi/gcp/types/input';

function camelToUnderscore(key: string): string {
  const result = key.replace(/([A-Z])/g, ' $1');
  return result.split(' ').join('_').toUpperCase();
}

export const config = new pulumi.Config();
export const stack = pulumi.getStack();
export const location = gcp.config.region || 'us-central1';

export const infra = new pulumi.StackReference(`idoshamun/infra/${stack}`);

export function createEnvVarsFromSecret(
  prefix: string,
): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv>[] {
  const envVars = config.requireObject<Record<string, string>>('env');
  return Object.keys(envVars).map(
    (key): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv> => {
      const secret = new gcp.secretmanager.Secret(`${prefix}-secret-${key}`, {
        secretId: `${prefix}-secret-${key}`,
        replication: { automatic: true },
      });

      const version = new gcp.secretmanager.SecretVersion(
        `${prefix}-sv-${key}`,
        {
          enabled: true,
          secret: secret.name,
          secretData: envVars[key],
        },
      );
      return {
        name: camelToUnderscore(key),
        value: pulumi
          .all([secret.secretId, version.id])
          .apply(
            ([name, version]) =>
              `gcp:///${name}/${version.split('/').reverse()[0]}`,
          ),
      };
    },
  );
}

export function serviceAccountToMember(
  serviceAccount: gcp.serviceaccount.Account,
): pulumi.Output<string> {
  return serviceAccount.email.apply((email) => `serviceAccount:${email}`);
}

export type IAMRole = { name: string; role: string };

export function addIAMRolesToServiceAccount(
  prefix: string,
  roles: IAMRole[],
  serviceAccount: gcp.serviceaccount.Account,
): void {
  const member = serviceAccountToMember(serviceAccount);
  roles.forEach(
    (role) =>
      new gcp.projects.IAMMember(`${prefix}-iam-${role.name}`, {
        role: role.role,
        member,
      }),
  );
}
