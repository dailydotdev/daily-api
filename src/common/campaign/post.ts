import { ValidationError } from 'apollo-server-errors';
import { AuthContext } from '../../Context';
import {
  ArticlePost,
  CampaignPost,
  CampaignState,
  CampaignType,
  Post,
  PostType,
  type ConnectionManager,
  type FreeformPost,
  type SharePost,
} from '../../entity';
import { getPostPermalink } from '../../schema/posts';
import {
  type GetCampaignListResponse,
  type GetCampaignResponse,
} from '../../integrations/skadi';
import type { Connection } from 'graphql-relay';
import { In } from 'typeorm';
import { mapCloudinaryUrl } from '../cloudinary';
import { pickImageUrl } from '../post';
import { NotFoundError } from '../../errors';
import { debeziumTimeToDate, systemUser, updateFlagsStatement } from '../utils';
import { getDiscussionLink } from '../links';
import { skadiApiClientV1 } from '../../integrations/skadi/api/v1/clients';
import { largeNumberFormat } from '../devcard';
import { formatMailDate, addNotificationEmailUtm } from '../mailing';
import { truncatePostToTweet } from '../twitter';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import { usdToCores } from '../number';

import {
  startCampaignTransferCores,
  stopCampaignTransferCores,
  type StartCampaignMutationArgs,
  type StopCampaignProps,
} from './common';
import { randomUUID } from 'crypto';
import {
  UserTransaction,
  UserTransactionProcessor,
  UserTransactionStatus,
  UserTransactionType,
} from '../../entity/user/UserTransaction';
import { addDays } from 'date-fns';
import { skadiApiClientV2 } from '../../integrations/skadi/api/v2/clients';

export interface GQLPromotedPost
  extends Omit<
    GetCampaignResponse,
    'spend' | 'budget' | 'startedAt' | 'endedAt'
  > {
  spend: number;
  budget: number;
  startedAt: Date;
  endedAt: Date;
}

export const validatePostBoostPermissions = async (
  ctx: AuthContext,
  postId: string,
): Promise<Pick<Post, 'id' | 'flags'>> => {
  const { userId } = ctx;

  return ctx.con.getRepository(Post).findOneOrFail({
    select: ['id', 'flags'],
    where: [
      { id: postId, authorId: userId },
      { id: postId, scoutId: userId },
    ],
  });
};

export const checkPostAlreadyBoosted = (post: Pick<Post, 'flags'>): void => {
  if (!!post.flags?.campaignId) {
    throw new ValidationError('Post is already boosted');
  }
};

interface CampaignBoostedPost
  extends Pick<Post, 'id' | 'shortId' | 'title' | 'slug'> {
  image: string;
  permalink: string;
  engagements: number;
  commentsPermalink?: string;
}

export interface GQLBoostedPost {
  campaign: GQLPromotedPost;
  post: CampaignBoostedPost;
}

interface GetBoostedPost extends CampaignBoostedPost {
  type: PostType;
  sharedTitle?: string;
  sharedImage?: string;
  views: number;
  upvotes: number;
  comments: number;
}

const getBoostedPostBuilder = (con: ConnectionManager, alias = 'p1') =>
  con
    .getRepository(Post)
    .createQueryBuilder(alias)
    .select(`"${alias}".id`, 'id')
    .addSelect(`"${alias}"."shortId"`, 'shortId')
    .addSelect(`"${alias}".slug`, 'slug')
    .addSelect(`"${alias}".image`, 'image')
    .addSelect(`"${alias}".title`, 'title')
    .addSelect(`"${alias}".type`, 'type')
    .addSelect(`"${alias}".upvotes::int`, 'upvotes')
    .addSelect(`"${alias}".comments::int`, 'comments')
    .addSelect(`"${alias}".views::int`, 'views')
    .addSelect('p2.title', 'sharedTitle')
    .addSelect('p2.image', 'sharedImage')
    .leftJoin(Post, 'p2', `"${alias}"."sharedPostId" = p2.id`);

export const getBoostedPost = async (
  con: ConnectionManager,
  id: string,
): Promise<GetBoostedPost> => {
  const result = await getBoostedPostBuilder(con).where({ id }).getRawOne();

  if (!result) {
    throw new NotFoundError('Post does not exist');
  }

  return result;
};

export const getFormattedBoostedPost = (
  post: GetBoostedPost,
): GQLBoostedPost['post'] => {
  const { id, shortId, sharedImage, sharedTitle, slug } = post;
  let image: string | undefined = post.image;
  let title = post.title;

  if (post.type === PostType.Share) {
    image = sharedImage;
    title = title || sharedTitle;
  }

  return {
    id,
    slug,
    shortId,
    title,
    image: mapCloudinaryUrl(image) ?? pickImageUrl({ createdAt: new Date() }),
    permalink: getPostPermalink({ shortId }),
    commentsPermalink: post.slug ? getDiscussionLink(post.slug) : undefined,
    engagements: post.comments + post.upvotes + post.views,
  };
};

export const getFormattedCampaign = ({
  spend,
  budget,
  startedAt,
  endedAt,
  ...campaign
}: GetCampaignResponse): GQLPromotedPost => ({
  ...campaign,
  spend: usdToCores(parseFloat(spend)),
  budget: usdToCores(parseFloat(budget)),
  startedAt: debeziumTimeToDate(startedAt),
  endedAt: debeziumTimeToDate(endedAt),
});

export interface BoostedPostStats
  extends Pick<GetCampaignListResponse, 'clicks' | 'impressions' | 'users'> {
  engagements: number;
  totalSpend: number;
}

export interface BoostedPostConnection extends Connection<GQLBoostedPost> {
  stats?: BoostedPostStats;
}

export const consolidateCampaignsWithPosts = async (
  campaigns: GetCampaignResponse[],
  con: ConnectionManager,
): Promise<GQLBoostedPost[]> => {
  const ids = campaigns.map(({ postId }) => postId);
  const builder = getBoostedPostBuilder(con);
  const postAlias = 'p1';
  const posts = await builder
    .where(`"${postAlias}".id IN (:...ids)`, { ids })
    .getRawMany<GetBoostedPost>();
  const mapped = posts.reduce(
    (map, post) => ({ ...map, [post.id]: post }),
    {} as Record<string, GetBoostedPost>,
  );

  return campaigns.map((campaign) => ({
    campaign: getFormattedCampaign(campaign),
    post: getFormattedBoostedPost(mapped[campaign.postId]),
  }));
};

interface TotalEngagements
  extends Pick<Post, 'views' | 'comments' | 'upvotes'> {}

export const getTotalEngagements = async (
  con: ConnectionManager,
  postIds: string[],
): Promise<number> => {
  const builder = con.getRepository(Post).createQueryBuilder();

  const engagements = await builder
    .select('SUM(upvotes)::int', 'upvotes')
    .addSelect('SUM(comments)::int', 'comments')
    .addSelect('SUM(views)::int', 'views')
    .where({ id: In(postIds) })
    .getRawOne<TotalEngagements>();

  return (
    (engagements?.upvotes || 0) +
    (engagements?.comments || 0) +
    (engagements?.views || 0)
  );
};

export const generateBoostEmailUpdate: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await skadiApiClientV1.getCampaignById({
    campaignId: notification.referenceId!,
    userId: user.id,
  });

  if (!campaign) {
    return null;
  }

  const post = await con.getRepository(Post).findOne({
    where: { id: campaign.postId },
  });

  if (!post) {
    return null;
  }

  const sharedPost = await (post.type === PostType.Share
    ? con.getRepository(ArticlePost).findOne({
        where: { id: (post as SharePost).sharedPostId },
        select: ['title', 'image', 'slug'],
      })
    : Promise.resolve(null));

  const title = truncatePostToTweet(post || sharedPost);
  const engagement = post.views + post.upvotes + post.comments;

  return {
    start_date: formatMailDate(debeziumTimeToDate(campaign.startedAt)),
    end_date: formatMailDate(debeziumTimeToDate(campaign.endedAt)),
    impressions: largeNumberFormat(campaign.impressions),
    clicks: largeNumberFormat(campaign.clicks),
    engagement: largeNumberFormat(engagement),
    post_link: getDiscussionLink(post.slug),
    analytics_link: addNotificationEmailUtm(
      notification.targetUrl,
      notification.type,
    ),
    post_image: sharedPost?.image || (post as FreeformPost).image,
    post_title: title,
  };
};

export const getAdjustedReach = (value: number) => {
  // We do plus-minus 8% of the generated value
  const difference = Math.floor(value * 0.08);
  const min = Math.max(value - difference, 0);
  const estimatedReach = {
    min,
    max: Math.max(value + difference, min),
  };

  return estimatedReach;
};

export const startCampaignPost = async (props: StartCampaignMutationArgs) => {
  const { ctx, args } = props;
  const { value: postId } = args;
  const post = await validatePostBoostPermissions(ctx, postId);
  checkPostAlreadyBoosted(post);

  const request = await ctx.con.transaction(async (manager) => {
    const id = randomUUID();
    const creativeId = randomUUID();
    const { budget, duration } = args;
    const total = budget * duration;
    const userId = ctx.userId;
    const endedAt = addDays(new Date(), duration);

    const campaign = await manager.getRepository(CampaignPost).save(
      manager.getRepository(CampaignPost).create({
        id,
        creativeId,
        flags: {
          budget: total,
          spend: 0,
          users: 0,
          clicks: 0,
          impressions: 0,
        },
        userId,
        referenceId: postId,
        state: CampaignState.Active,
        endedAt,
        postId,
        type: CampaignType.Post,
      }),
    );

    const campaignId = campaign.id;
    const tags = await getPostTags(manager, postId);

    await skadiApiClientV2.startCampaign(campaign, tags);

    await manager
      .getRepository(Post)
      .update(
        { id: postId },
        { flags: updateFlagsStatement<Post>({ campaignId: id }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: systemUser.id,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: userId,
        value: total,
        valueIncFees: 0,
        fee: 0,
        request: ctx.requestMeta,
        flags: { note: `Post Boost started` },
        referenceId: campaignId,
        referenceType: UserTransactionType.PostBoost,
      }),
    );

    return await startCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return request.transaction;
};

export const stopCampaignPost = async ({
  campaign,
  ctx,
}: StopCampaignProps) => {
  const { id: campaignId, userId, referenceId } = campaign;

  const { budget } = await skadiApiClientV2.cancelCampaign({
    campaignId,
    userId,
  });

  const result = await ctx.con.transaction(async (manager) => {
    const toRefund = parseFloat(budget);

    await manager
      .getRepository(Post)
      .update(
        { id: referenceId },
        { flags: updateFlagsStatement<Post>({ campaignId: null }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: userId,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: systemUser.id,
        value: usdToCores(toRefund),
        valueIncFees: 0,
        fee: 0,
        flags: { note: `Post Boost refund` },
        referenceId: campaignId,
        referenceType: UserTransactionType.PostBoost,
      }),
    );

    return await stopCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return result.transaction;
};

export const getPostTags = async (con: ConnectionManager, postId: string) => {
  const builder = con.getRepository(Post).createQueryBuilder('p1');
  const subquery = builder
    .subQuery()
    .select(`COALESCE(p2."tagsStr", '')`)
    .from(Post, 'p2')
    .where('p2.id = p1."sharedPostId"')
    .getQuery();

  const result = await builder
    .select(`COALESCE(p1."tagsStr", '')`, 'tagsStr')
    .addSelect(`(${subquery})`, 'sharedTagsStr')
    .where('p1.id = :id', { id: postId })
    .getRawOne<{ tagsStr: string; sharedTagsStr: string }>();

  const tags1 = (result?.tagsStr ?? '').split(',');
  const tags2 = (result?.sharedTagsStr ?? '').split(',');
  const list = [...tags1, ...tags2].filter((tag) => tag.trim().length > 0);

  return Array.from(new Set(list));
};
