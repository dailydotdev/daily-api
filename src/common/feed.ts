import { DataSource } from 'typeorm';
import { Feed } from '../entity/Feed';
import { ValidationError } from 'apollo-server-errors';
import { SubmissionFailErrorMessage } from '../errors';

export const getFeedByIdentifiersOrFail = async ({
  con,
  feedIdOrSlug,
  userId,
}: {
  con: DataSource;
  feedIdOrSlug: string;
  userId: string;
}): Promise<Feed | never> => {
  const feed = await con.getRepository(Feed).findOneOrFail({
    where: [
      {
        id: feedIdOrSlug,
        userId: userId,
      },
      {
        slug: feedIdOrSlug,
        userId: userId,
      },
    ],
  });

  return feed;
};

export const maxFeedNameLength = 50;

export const feedNameMatcher = /^[a-z0-9 ]+$/i;

export const validateFeedPayload = ({
  name,
}: {
  name: Feed['flags']['name'];
}): never | undefined => {
  if (!name) {
    throw new ValidationError(SubmissionFailErrorMessage.FEED_NAME_REQUIRED);
  }

  if (name.length > maxFeedNameLength) {
    throw new ValidationError(SubmissionFailErrorMessage.FEED_NAME_LENGTH);
  }

  if (!feedNameMatcher.test(name)) {
    throw new ValidationError(SubmissionFailErrorMessage.FEED_NAME_INVALID);
  }
};
