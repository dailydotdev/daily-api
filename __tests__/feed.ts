import { FastifyInstance } from 'fastify';
import { Connection, getConnection } from 'typeorm';
import { ApolloServer } from 'apollo-server-fastify';
import {
  ApolloServerTestClient,
  createTestClient,
} from 'apollo-server-testing';
import * as request from 'supertest';
import * as _ from 'lodash';

import createApolloServer from '../src/apollo';
import { Context } from '../src/Context';
import { MockContext, saveFixtures } from './helpers';
import appFunc from '../src';
import { Post, PostTag, Source, SourceDisplay } from '../src/entity';
import { sourcesFixture } from './fixture/source';
import { postsFixture, postTagsFixture } from './fixture/post';
import { sourceDisplaysFixture } from './fixture/sourceDisplay';
import { Ranking } from '../src/schema/feed';

let app: FastifyInstance;
let con: Connection;
let server: ApolloServer;
let client: ApolloServerTestClient;
let loggedUser: string = null;

beforeAll(async () => {
  con = await getConnection();
  server = await createApolloServer({
    context: (): Context => new MockContext(con, loggedUser),
    playground: false,
  });
  client = createTestClient(server);
  app = await appFunc();
  return app.ready();
});

beforeEach(async () => {
  loggedUser = null;

  await saveFixtures(con, Source, sourcesFixture);
  await saveFixtures(con, SourceDisplay, sourceDisplaysFixture);
  await saveFixtures(con, Post, postsFixture);
  await saveFixtures(con, PostTag, postTagsFixture);
});

afterAll(() => app.close());

describe('query anonymousFeed', () => {
  const QUERY = (
    ranking: Ranking = Ranking.POPULARITY,
    now = new Date(),
    first = 10,
  ): string => `
  query AnonymousFeed($filters: FiltersInput) {
    anonymousFeed(filters: $filters, ranking: ${ranking}, now: "${now.toISOString()}", first: ${first}) {
      pageInfo {
        endCursor
        hasNextPage
      }
      edges {
        node {
          id
          url
          title
          image
          ratio
          placeholder
          readTime
          tags
          source {
            id
            name
            image
            public
          }
        }
      }
    }
  }
`;

  it('should return anonymous feed with no filters ordered by popularity', async () => {
    const res = await client.query({ query: QUERY() });
    expect(res.data).toMatchSnapshot();
  });

  it('should return anonymous feed with no filters ordered by time', async () => {
    const res = await client.query({ query: QUERY(Ranking.TIME) });
    expect(res.data).toMatchSnapshot();
  });

  it('should return anonymous feed filtered by sources', async () => {
    const res = await client.query({
      query: QUERY(),
      variables: { filters: { includeSources: ['a', 'b'] } },
    });
    expect(res.data).toMatchSnapshot();
  });

  it('should return anonymous feed filtered by tags', async () => {
    const res = await client.query({
      query: QUERY(),
      variables: { filters: { includeTags: ['html', 'webdev'] } },
    });
    expect(res.data).toMatchSnapshot();
  });

  it('should return anonymous feed while excluding sources', async () => {
    const res = await client.query({
      query: QUERY(),
      variables: { filters: { excludeSources: ['a'] } },
    });
    expect(res.data).toMatchSnapshot();
  });

  it('should return anonymous feed filtered by tags and sources', async () => {
    const res = await client.query({
      query: QUERY(),
      variables: {
        filters: {
          includeTags: ['javascript'],
          includeSources: ['a', 'b'],
        },
      },
    });
    expect(res.data).toMatchSnapshot();
  });
});

describe('compatibility routes', () => {
  describe('GET /posts/latest', () => {
    it('should return anonymous feed with no filters ordered by popularity', async () => {
      const res = await request(app.server)
        .get('/v1/posts/latest')
        .query({ latest: new Date(), pageSize: 2, page: 0 })
        .send()
        .expect(200);
      expect(res.body.map((x) => _.omit(x, ['createdAt']))).toMatchSnapshot();
    });

    it('should return anonymous feed filtered by sources', async () => {
      const res = await request(app.server)
        .get('/v1/posts/latest')
        .query({ latest: new Date(), sources: ['a', 'b'] })
        .send()
        .expect(200);
      expect(res.body.map((x) => _.omit(x, ['createdAt']))).toMatchSnapshot();
    });

    it('should return anonymous feed filtered by tags', async () => {
      const res = await request(app.server)
        .get('/v1/posts/latest')
        .query({ latest: new Date(), tags: ['html', 'webdev'] })
        .send()
        .expect(200);
      expect(res.body.map((x) => _.omit(x, ['createdAt']))).toMatchSnapshot();
    });

    it('should return anonymous feed filtered by tags and sources', async () => {
      const res = await request(app.server)
        .get('/v1/posts/latest')
        .query({
          latest: new Date(),
          tags: ['javascript'],
          sources: ['a', 'b'],
        })
        .send()
        .expect(200);
      expect(res.body.map((x) => _.omit(x, ['createdAt']))).toMatchSnapshot();
    });
  });
});
