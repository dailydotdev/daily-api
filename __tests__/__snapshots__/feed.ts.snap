// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compatibility routes GET /posts/latest should return anonymous feed filtered by sources 1`] = `
Array [
  Object {
    "id": "p5",
    "image": null,
    "placeholder": null,
    "publishedAt": null,
    "ratio": null,
    "readTime": null,
    "tags": Array [
      "html",
      "javascript",
    ],
    "title": "P5",
    "url": "http://p5.com",
  },
  Object {
    "id": "p2",
    "image": null,
    "placeholder": null,
    "publishedAt": null,
    "ratio": null,
    "readTime": null,
    "tags": Array [],
    "title": "P2",
    "url": "http://p2.com",
  },
  Object {
    "id": "p4",
    "image": null,
    "placeholder": null,
    "publishedAt": null,
    "ratio": null,
    "readTime": null,
    "tags": Array [
      "backend",
      "data",
      "javascript",
    ],
    "title": "P4",
    "url": "http://p4.com",
  },
  Object {
    "id": "p1",
    "image": null,
    "placeholder": null,
    "publishedAt": null,
    "ratio": null,
    "readTime": null,
    "tags": Array [
      "javascript",
      "webdev",
    ],
    "title": "P1",
    "url": "http://p1.com",
  },
]
`;

exports[`compatibility routes GET /posts/latest should return anonymous feed filtered by tags 1`] = `
Array [
  Object {
    "id": "p5",
    "image": null,
    "placeholder": null,
    "publishedAt": null,
    "ratio": null,
    "readTime": null,
    "tags": Array [
      "html",
      "javascript",
    ],
    "title": "P5",
    "url": "http://p5.com",
  },
  Object {
    "id": "p1",
    "image": null,
    "placeholder": null,
    "publishedAt": null,
    "ratio": null,
    "readTime": null,
    "tags": Array [
      "javascript",
      "webdev",
    ],
    "title": "P1",
    "url": "http://p1.com",
  },
]
`;

exports[`compatibility routes GET /posts/latest should return anonymous feed filtered by tags and sources 1`] = `
Array [
  Object {
    "id": "p5",
    "image": null,
    "placeholder": null,
    "publishedAt": null,
    "ratio": null,
    "readTime": null,
    "tags": Array [
      "html",
      "javascript",
    ],
    "title": "P5",
    "url": "http://p5.com",
  },
  Object {
    "id": "p4",
    "image": null,
    "placeholder": null,
    "publishedAt": null,
    "ratio": null,
    "readTime": null,
    "tags": Array [
      "backend",
      "data",
      "javascript",
    ],
    "title": "P4",
    "url": "http://p4.com",
  },
  Object {
    "id": "p1",
    "image": null,
    "placeholder": null,
    "publishedAt": null,
    "ratio": null,
    "readTime": null,
    "tags": Array [
      "javascript",
      "webdev",
    ],
    "title": "P1",
    "url": "http://p1.com",
  },
]
`;

exports[`compatibility routes GET /posts/latest should return anonymous feed with no filters ordered by popularity 1`] = `
Array [
  Object {
    "id": "p5",
    "image": null,
    "placeholder": null,
    "publishedAt": null,
    "ratio": null,
    "readTime": null,
    "tags": Array [
      "html",
      "javascript",
    ],
    "title": "P5",
    "url": "http://p5.com",
  },
  Object {
    "id": "p2",
    "image": null,
    "placeholder": null,
    "publishedAt": null,
    "ratio": null,
    "readTime": null,
    "tags": Array [],
    "title": "P2",
    "url": "http://p2.com",
  },
]
`;

exports[`query anonymousFeed should return anonymous feed filtered by sources 1`] = `
Object {
  "anonymousFeed": Object {
    "edges": Array [
      Object {
        "node": Object {
          "id": "p5",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "b",
            "image": "http://image.com/b",
            "name": "B",
            "public": true,
          },
          "tags": Array [
            "html",
            "javascript",
          ],
          "title": "P5",
          "url": "http://p5.com",
        },
      },
      Object {
        "node": Object {
          "id": "p2",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "b",
            "image": "http://image.com/b",
            "name": "B",
            "public": true,
          },
          "tags": Array [],
          "title": "P2",
          "url": "http://p2.com",
        },
      },
      Object {
        "node": Object {
          "id": "p4",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "a",
            "image": "http://image.com/a",
            "name": "A",
            "public": true,
          },
          "tags": Array [
            "backend",
            "data",
            "javascript",
          ],
          "title": "P4",
          "url": "http://p4.com",
        },
      },
      Object {
        "node": Object {
          "id": "p1",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "a",
            "image": "http://image.com/a",
            "name": "A",
            "public": true,
          },
          "tags": Array [
            "javascript",
            "webdev",
          ],
          "title": "P1",
          "url": "http://p1.com",
        },
      },
    ],
    "pageInfo": Object {
      "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "hasNextPage": false,
    },
  },
}
`;

exports[`query anonymousFeed should return anonymous feed filtered by tags 1`] = `
Object {
  "anonymousFeed": Object {
    "edges": Array [
      Object {
        "node": Object {
          "id": "p5",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "b",
            "image": "http://image.com/b",
            "name": "B",
            "public": true,
          },
          "tags": Array [
            "html",
            "javascript",
          ],
          "title": "P5",
          "url": "http://p5.com",
        },
      },
      Object {
        "node": Object {
          "id": "p1",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "a",
            "image": "http://image.com/a",
            "name": "A",
            "public": true,
          },
          "tags": Array [
            "javascript",
            "webdev",
          ],
          "title": "P1",
          "url": "http://p1.com",
        },
      },
    ],
    "pageInfo": Object {
      "endCursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "hasNextPage": false,
    },
  },
}
`;

exports[`query anonymousFeed should return anonymous feed filtered by tags and sources 1`] = `
Object {
  "anonymousFeed": Object {
    "edges": Array [
      Object {
        "node": Object {
          "id": "p5",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "b",
            "image": "http://image.com/b",
            "name": "B",
            "public": true,
          },
          "tags": Array [
            "html",
            "javascript",
          ],
          "title": "P5",
          "url": "http://p5.com",
        },
      },
      Object {
        "node": Object {
          "id": "p4",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "a",
            "image": "http://image.com/a",
            "name": "A",
            "public": true,
          },
          "tags": Array [
            "backend",
            "data",
            "javascript",
          ],
          "title": "P4",
          "url": "http://p4.com",
        },
      },
      Object {
        "node": Object {
          "id": "p1",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "a",
            "image": "http://image.com/a",
            "name": "A",
            "public": true,
          },
          "tags": Array [
            "javascript",
            "webdev",
          ],
          "title": "P1",
          "url": "http://p1.com",
        },
      },
    ],
    "pageInfo": Object {
      "endCursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "hasNextPage": false,
    },
  },
}
`;

exports[`query anonymousFeed should return anonymous feed while excluding sources 1`] = `
Object {
  "anonymousFeed": Object {
    "edges": Array [
      Object {
        "node": Object {
          "id": "p5",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "b",
            "image": "http://image.com/b",
            "name": "B",
            "public": true,
          },
          "tags": Array [
            "html",
            "javascript",
          ],
          "title": "P5",
          "url": "http://p5.com",
        },
      },
      Object {
        "node": Object {
          "id": "p2",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "b",
            "image": "http://image.com/b",
            "name": "B",
            "public": true,
          },
          "tags": Array [],
          "title": "P2",
          "url": "http://p2.com",
        },
      },
      Object {
        "node": Object {
          "id": "p3",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "c",
            "image": "http://image.com/c",
            "name": "C",
            "public": true,
          },
          "tags": Array [],
          "title": "P3",
          "url": "http://p3.com",
        },
      },
    ],
    "pageInfo": Object {
      "endCursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "hasNextPage": false,
    },
  },
}
`;

exports[`query anonymousFeed should return anonymous feed with no filters ordered by popularity 1`] = `
Object {
  "anonymousFeed": Object {
    "edges": Array [
      Object {
        "node": Object {
          "id": "p5",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "b",
            "image": "http://image.com/b",
            "name": "B",
            "public": true,
          },
          "tags": Array [
            "html",
            "javascript",
          ],
          "title": "P5",
          "url": "http://p5.com",
        },
      },
      Object {
        "node": Object {
          "id": "p2",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "b",
            "image": "http://image.com/b",
            "name": "B",
            "public": true,
          },
          "tags": Array [],
          "title": "P2",
          "url": "http://p2.com",
        },
      },
      Object {
        "node": Object {
          "id": "p3",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "c",
            "image": "http://image.com/c",
            "name": "C",
            "public": true,
          },
          "tags": Array [],
          "title": "P3",
          "url": "http://p3.com",
        },
      },
      Object {
        "node": Object {
          "id": "p4",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "a",
            "image": "http://image.com/a",
            "name": "A",
            "public": true,
          },
          "tags": Array [
            "backend",
            "data",
            "javascript",
          ],
          "title": "P4",
          "url": "http://p4.com",
        },
      },
      Object {
        "node": Object {
          "id": "p1",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "a",
            "image": "http://image.com/a",
            "name": "A",
            "public": true,
          },
          "tags": Array [
            "javascript",
            "webdev",
          ],
          "title": "P1",
          "url": "http://p1.com",
        },
      },
    ],
    "pageInfo": Object {
      "endCursor": "YXJyYXljb25uZWN0aW9uOjQ=",
      "hasNextPage": false,
    },
  },
}
`;

exports[`query anonymousFeed should return anonymous feed with no filters ordered by time 1`] = `
Object {
  "anonymousFeed": Object {
    "edges": Array [
      Object {
        "node": Object {
          "id": "p1",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "a",
            "image": "http://image.com/a",
            "name": "A",
            "public": true,
          },
          "tags": Array [
            "javascript",
            "webdev",
          ],
          "title": "P1",
          "url": "http://p1.com",
        },
      },
      Object {
        "node": Object {
          "id": "p2",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "b",
            "image": "http://image.com/b",
            "name": "B",
            "public": true,
          },
          "tags": Array [],
          "title": "P2",
          "url": "http://p2.com",
        },
      },
      Object {
        "node": Object {
          "id": "p3",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "c",
            "image": "http://image.com/c",
            "name": "C",
            "public": true,
          },
          "tags": Array [],
          "title": "P3",
          "url": "http://p3.com",
        },
      },
      Object {
        "node": Object {
          "id": "p4",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "a",
            "image": "http://image.com/a",
            "name": "A",
            "public": true,
          },
          "tags": Array [
            "backend",
            "data",
            "javascript",
          ],
          "title": "P4",
          "url": "http://p4.com",
        },
      },
      Object {
        "node": Object {
          "id": "p5",
          "image": null,
          "placeholder": null,
          "ratio": null,
          "readTime": null,
          "source": Object {
            "id": "b",
            "image": "http://image.com/b",
            "name": "B",
            "public": true,
          },
          "tags": Array [
            "html",
            "javascript",
          ],
          "title": "P5",
          "url": "http://p5.com",
        },
      },
    ],
    "pageInfo": Object {
      "endCursor": "YXJyYXljb25uZWN0aW9uOjQ=",
      "hasNextPage": false,
    },
  },
}
`;
